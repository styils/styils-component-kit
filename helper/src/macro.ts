import { createMacro } from 'babel-plugin-macros'
import * as t from '@babel/types'
import { addNamed } from '@babel/helper-module-imports'
import type { NodePath } from '@babel/traverse'
import type { MParams } from './types'
import { component } from './component'
import { state } from './state'
import { proper } from './proper'
import { useMount } from './useMount'
import { useMemo } from './useMemo'
import { useWatchEffect } from './useWatchEffect'

/**
 * Avoid importing multiple identical methods
 */
const hookCacheId = new Map()

export default createMacro(({ references, state: babelState }) => {
  /**
   * used to get all props states
   * Used to get all props state, auto-populate react deps
   */
  const variableMaps = new Set([])
  const {
    component: Icomponent = [],
    proper: Iproper = [],
    state: Istate = [],
    ...macros
  } = references

  const createAddImportPath = (path: NodePath) => (name: string, source: string) => {
    const cacheCode = babelState.file.code + name

    const nameId = hookCacheId.get(cacheCode) ?? addNamed(path, name, source)
    hookCacheId.set(cacheCode, nameId)

    return nameId
  }

  // component always stays first
  Icomponent.forEach((path) => {
    component(path, (babelState.opts as { frame: string }).frame, createAddImportPath(path))
  })

  Iproper.forEach((path) => {
    proper(path, { addImportName: createAddImportPath(path), ...babelState }, variableMaps)
  })

  Istate.forEach((path) => {
    state(path, { addImportName: createAddImportPath(path), ...babelState }, variableMaps)
  })

  const macroHookMethods = [useMount, useMemo, useWatchEffect]

  macroHookMethods.forEach((macro) => {
    macros[macro.name]?.forEach((path) => {
      if (!t.isCallExpression(path.parentPath.node)) {
        // Macros must be called directly, otherwise the correct call cannot be traced
        throw new Error('All macros must be called')
      }

      const currentCallExpression = path.findParent((p) =>
        t.isCallExpression(p)
      ) as MParams['currentCallExpression']

      macro(
        path,
        {
          currentCallExpression,
          ...babelState, // `import .. form ..` helper method
          addImportName: createAddImportPath(path)
        },
        variableMaps
      )
    })
  })
})
