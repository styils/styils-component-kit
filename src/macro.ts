import { createMacro } from 'babel-plugin-macros'
import { state as IState } from './state'
import { proper as IProper } from './proper'

export declare function state<T>(initialState: T): T

export default createMacro(({ references, state }) => {
  if (references.state) references.state.forEach((path) => IState(path, state))

  if (references.proper) references.proper.forEach((path) => IProper(path, state))
})
