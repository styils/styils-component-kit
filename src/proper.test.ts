import { getCode } from '../tests'

describe('use props', () => {
  it('base react', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props)

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'react' }
    )

    expect(code).toMatchSnapshot()
  })

  it('base vue', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo, bar} = proper(props)

      return (
        <button>
          Clicked {foo} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'vue' }
    )

    expect(code).toMatchSnapshot()
  })

  it('base solid', () => {
    const code = getCode(
      `
    import { proper } from 'macro'

    export default (props) => {
      const {foo:foo1, bar, ...rest} = proper(props)

      return (
        <button {...rest}>
          Clicked {foo1} {bar === 1 ? 'time' : 'times'}
        </button>
      )
    }
    `,
      { frame: 'solid' }
    )

    expect(code).toMatchSnapshot()
  })
})
