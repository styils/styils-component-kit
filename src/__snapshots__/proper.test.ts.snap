// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use props base react 1`] = `
"export default (props) => {
  const { foo, bar } = props;
  return (
    <button>
      Clicked {foo} {bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props base solid 1`] = `
"import { splitProps as _splitProps } from "solid-js";
export default (props) => {
  const [_local, rest] = _splitProps(props, ["foo", "bar"]);

  return (
    <button {...rest}>
      Clicked {_local.foo} {_local.bar === 1 ? "time" : "times"}
    </button>
  );
};
"
`;

exports[`use props base vue 1`] = `
"import { toRefs as _toRefs } from "vue";
export default (props) => {
  const { foo, bar } = _toRefs(props);

  return (
    <button>
      Clicked {foo.value} {bar.value === 1 ? "time" : "times"}
    </button>
  );
};
"
`;
